### ISSUE TITLE
Big difference in output between exllama1_hf and exllama2_hf

### ISSUE THREAD

**Issue by BadisG** [turboderp/exllamav2#14]
Hello,
I noticed the quality of the output decreased with exllama2 so I took a look at the logits, it's the same model, same quant, same samplers, same prompt, same seed
![YdSGDl2](https://github.com/turboderp/exllamav2/assets/110173477/317405bf-f876-44e4-a1b8-230921e98ab4)
Maybe it's a bug on ooba's webui I don't know

**Comment by BadisG** [comment]
To support my claim I did a perplexity test and yeah... the perplexity is worse in exllama2
![zrz](https://github.com/turboderp/exllamav2/assets/110173477/8696c280-bb14-4b45-b13f-1556a4775e73)

**Comment by Thireus** [comment]
Something I've noticed as well.

**Comment by turboderp** [comment]
I've made some changes to the GPTQ kernel to increase precision. They're in the test branch for now, since I need to confirm that they don't break anything (on ROCm in particular).

**Comment by BadisG** [comment]
Will this also improve precision on exl2 models?

**Comment by turboderp** [comment]
Somewhat, but the changes there are different. EXL2 still gets better ppl than GPTQ for the same bitrate, and I'll do some more experiments to quantify it all soon. Currently I'm seeing how far I can push the Llama-2 70B quants. So far it remains coherent down to 2.35 bits, and the 2.3 bit quant is pending.

**Comment by BadisG** [comment]
Yeah it gives better ppl than GPTQ but that's because of the quality of your quantization method, not really because of the loader, if the exllama2 loader gave bad results overall on GPTQ, it means it also did some damage on the ppl of exl2 models. It also means that it's possible to increase the precision even further on the exl2 models and get even better perplexity overall, but that's just my 2 cents and maybe it was doing wrong only on the GPTQ models.

**Comment by BadisG** [comment]
All right I've made some new measurements with your branch "test" and here's the results

![fsfs](https://github.com/turboderp/exllamav2/assets/110173477/6b09c650-752c-4bdb-9f53-5b35fde69b68)

- For GPTQ models: it's better but still not as precise as exllama1
- For exl2 models: nothing changed

**Comment by turboderp** [comment]
There's likely something wrong with the way you're measuring perplexity, or an issue with the ooba plugin. If I port the dataset loader and perplexity test over from V2 to V1 to make sure they're actually testing on the same data with the same cross-entropy calculation, I get this:

ExLlama:

```
(q4) [bb@bbc exllama]$ python _test_pplv2.py -m /mnt/str/models/_test_models/TheBloke_Llama-2-7B-GPTQ/32g-act/ -ed /mnt/str/datasets/wikitext-test.parquet 
 -- Running perplexity test
 -- Dataset: /mnt/str/datasets/wikitext-test.parquet
 -- Tokenizing eval data, 128 rows x 2048 tokens...
 -- First 50 tokens of dataset:
    'Robert Boulter is an English film , television and theatre actor . He had a guest @-@ starin role on the television series The Bill in 2000 . This was followed by a starring role in the play Herons'
 -- Last 50 tokens of dataset:
    'aged man who has worked hard since his youth . He cares deeply for his motherless daughter , Mary , but was unable to attend da annual commencement at her co @-@ educational college . He awaits fir her to return from'
 -- Inference.............
 -- Evaluation perplexity: 9.0059
```

ExLlamaV2:

```
(q4) [bb@bbc exllamav2]$ python test_inference.py -m /mnt/str/models/_test_models/TheBloke_Llama-2-7B-GPTQ/32g-act/ -ed /mnt/str/datasets/wikitext-test.parquet 
 -- Model: /mnt/str/models/_test_models/TheBloke_Llama-2-7B-GPTQ/32g-act/
 -- Options: ['rope_scale 1.0', 'rope_alpha 1.0']
 -- Loading model...
 -- Loading tokenizer...
 -- Running perplexity test
 -- Dataset: /mnt/str/datasets/wikitext-test.parquet
 -- Tokenizing eval data, 128 rows x 2048 tokens...
 -- First 50 tokens of dataset:
    'Robert Boulter is an English film , television and theatre actor . He had a guest @-@ starin role on the television series The Bill in 2000 . This was followed by a starring role in the play Herons'
 -- Last 50 tokens of dataset:
    'aged man who has worked hard since his youth . He cares deeply for his motherless daughter , Mary , but was unable to attend da annual commencement at her co @-@ educational college . He awaits fir her to return from'
 -- Inference.............
 -- Evaluation perplexity: 9.0053
(q4) [bb@bbc exllamav2]$ 
```

**Comment by BadisG** [comment]
try it with the wikitest and see if you have also the same scores as me, it's the one we always use to measure perplexity

**Comment by turboderp** [comment]
That was wikitest

**Comment by BadisG** [comment]
Oh... then maybe the difference isn't visible with 7b models? Try it with mythomax 13b like I did and see if the results are similar, this way we could compare apples to apples

**Comment by ardfork** [comment]
> I've made some changes to the GPTQ kernel to increase precision. They're in the test branch for now, since I need to confirm that they don't break anything (on ROCm in particular).

Works fine on my machine. FYI you can replace `__half2float(__low2half(x))` by `__low2float(x)`, `__half2half2(__low2half(x))` by `__low2half2(x)` and same for the high part.
